src/bluetooth_linux.c

```diff
@@ -0,0 +1,673 @@
+					DBUS_CHAR_IFACE, "ReadValue", NULL, &reply, "a{sv}", 0))
+			goto out;
+
+		if (0 > sd_bus_message_read(reply, "ay", 2, cp_len, cp_len + 1))
+			goto out;
+
+		sd_bus_message_unref(reply);
+		reply = NULL;
+
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.service_revision,
+					DBUS_CHAR_IFACE, "ReadValue", NULL, &reply, "a{sv}", 0))
+			goto out;
+		if (0 > sd_bus_message_read(reply, "ay", 1, &revision))
+			goto out;
+
+		/* for simplicity, we allow now only FIDO2 */
+		if (!(revision & 0x20))
+			goto out;
+
+		newdev->controlpoint_size = ((size_t)cp_len[0] << 8) + cp_len[1];
+		newdev->reply_buf = calloc(1, newdev->controlpoint_size);
+
+		if (0 > sd_bus_call_method(newdev->bus, "org.bluez", newdev->paths.status,
+					DBUS_CHAR_IFACE, "StartNotify", NULL, NULL, ""))
+			goto out;
+
+		if (0 > sd_bus_match_signal(newdev->bus, &newdev->slot, "org.bluez",
+					newdev->paths.status,
+					"org.freedesktop.DBus.Properties", "PropertiesChanged",
+					read_cb, newdev))
+			goto out;
+
+		return newdev;
+	}
+out:
+	if (reply)
+		sd_bus_message_unref(reply);
+
+	if (newdev->reply_buf)
+		explicit_bzero(newdev->reply_buf, newdev->controlpoint_size);
+	free(newdev->reply_buf);
+	free(newdev->paths.service_revision);
+	free(newdev->paths.control_point_length);
+	free(newdev->paths.control_point);
+	free(newdev->paths.service);
+	free(newdev->paths.dev);
+	if (newdev->slot)
+		sd_bus_slot_unref(newdev->slot);
+
+	if (newdev->bus)
+		sd_bus_unref(newdev->bus);
+
+	free(newdev);
+	return NULL;
+}
+
+void fido_bluetooth_close(void *handle)
+{
+	struct ble *dev = (struct ble *)handle;
+	sd_bus_call_method(dev->bus, "org.bluez", dev->paths.status,
+			   DBUS_CHAR_IFACE, "StopNotify", NULL, NULL, "");
+	if (dev->reply_buf)
+		explicit_bzero(dev->reply_buf, dev->controlpoint_size);
+	free(dev->reply_buf);
+	free(dev->paths.service_revision);
+	free(dev->paths.control_point_length);
+	free(dev->paths.control_point);
+	free(dev->paths.service);
+	free(dev->paths.dev);
+	if (dev->slot)
+		sd_bus_slot_unref(dev->slot);
+
+	if (dev->bus)
+		sd_bus_unref(dev->bus);
+
+	free(dev);
+}
+
+int
+fido_bluetooth_read(void *handle, unsigned char *buf, size_t len, int ms)
+{
+	struct ble *dev = (struct ble *)handle;
+	dev->reply_len = 0;
+	while(dev->reply_len == 0) {
+		int ret = sd_bus_process(dev->bus, NULL);
+		if (ret < 0)
+			return FIDO_ERR_INTERNAL;
+
+		if (ret == 0) {
+			ret = sd_bus_wait(dev->bus, ms < 0 ? UINT64_MAX : (uint64_t)ms * 1000);
+			if (ret == 0) /* timeout */
+				return -1;
+
+			if (ret < 0)
+				return FIDO_ERR_INTERNAL;
+		} else if (dev->reply_len > 0) {
+			if (dev->reply_len < len)
+				len = dev->reply_len;
+
+			dev->reply_len = 0;
+			memcpy(buf, dev->reply_buf, len);
+			return (int)len;
+		}
+	}
+
+	return FIDO_ERR_INTERNAL;
+}
+
+int
+fido_bluetooth_write(void *handle, const unsigned char *buf, size_t len)
+{
+	struct ble *dev = (struct ble *)handle;
+	sd_bus_message *send_msg;
```

Comment:
I believe a `sd_bus_message_unref` is needed here?

Reply:
yes, I will fix it